algebra/Atiyah.lean

```diff
@@ -0,0 +1,1695 @@
+             is_ideal.mul_mem $ subset_generate _ ⟨m, ny, by simp [*]⟩,
+             is_ideal.mul_mem $ subset_generate _ ⟨m, nz, by simp [*]⟩,
+             by rw [← hyz, hymn, hnyz, mul_add, add_mul]⟩)
+         (ih (list.sum tl) (λ z hz, hx z (or.inr hz)) rfl),
+            by simp at hLx; rwa hLx at h))
+   x hx hLx,
+ λ hx, mul_add_subset hx⟩
+
+theorem add_congr : S₁ = S₂ → S₃ = S₄ → S₁ + S₃ = S₂ + S₄ :=
+λ h12 h34, calc
+  S₁ + S₃ = S₂ + S₃ : @eq.subst _ (λ T, ∀ H:is_ideal α T, S₁ + S₃ = @@add _ T S₃ H _) S₁ S₂ h12 (λ H, rfl) _
+      ... = S₂ + S₄ : @eq.subst _ (λ T, ∀ H:is_ideal α T, S₂ + S₃ = @@add _ S₂ T _ H) S₃ S₄ h34 (λ H, rfl) _
+
+theorem add_mul : (S₁ + S₂) * S₃ = (S₁ * S₃) + (S₂ * S₃) :=
+calc (S₁ + S₂) * S₃
+      = S₃ * (S₁ + S₂) : mul_comm
+  ... = (S₃ * S₁) + (S₃ * S₂) : mul_add
+  ... = (S₁ * S₃) + (S₂ * S₃) : add_congr mul_comm mul_comm
+
+theorem mul_congr : S₁ = S₂ → S₃ = S₄ → S₁ * S₃ = S₂ * S₄ :=
+λ h12 h34, calc
+  S₁ * S₃ = S₂ * S₃ : @eq.subst _ (λ T, ∀ H:is_ideal α T, S₁ * S₃ = @@mul _ T S₃ H _) S₁ S₂ h12 (λ H, rfl) _
+      ... = S₂ * S₄ : @eq.subst _ (λ T, ∀ H:is_ideal α T, S₂ * S₃ = @@mul _ S₂ T _ H) S₃ S₄ h34 (λ H, rfl) _
+
+theorem mul_subset_of_subset : S₂ ⊆ S₃ → S₁ * S₂ ⊆ S₁ * S₃ :=
+λ h x hx S s hs, @@hx S s $ λ z ⟨p, q, hp, hq, hz⟩, hs ⟨p, q, hp, h hq, hz⟩
+
+theorem mul_subset_of_subset_of_subset : S₁ ⊆ S₂ → S₃ ⊆ S₄ → S₁ * S₃ ⊆ S₂ * S₄ :=
+λ h12 h34 x hx S s hs, @@hx S s $ λ z ⟨p, q, hp, hq, hz⟩, hs ⟨p, q, h12 hp, h34 hq, hz⟩
+
+theorem add_subset : S₁ ⊆ S₃ → S₂ ⊆ S₃ → S₁ + S₂ ⊆ S₃ :=
+λ h13 h23 x ⟨y, z, hy, hz, hx⟩, set.mem_of_eq_of_mem hx
+  $ add_mem (h13 hy) (h23 hz)
+
+theorem add_subset_of_subset_of_subset : S₁ ⊆ S₂ → S₃ ⊆ S₄ → S₁ + S₃ ⊆ S₂ + S₄ :=
+λ h12 h34 x ⟨y, z, hy, hz, hx⟩, set.mem_of_eq_of_mem hx
+  $ add_mem (subset_add_left $ h12 hy) (subset_add_right $ h34 hz)
+
+theorem add_mul_inter_subset_mul : (S₁ + S₂) * (S₁ ∩ S₂) ⊆ S₁ * S₂ :=
+calc
+  (S₁ + S₂) * (S₁ ∩ S₂) = S₁ * (S₁ ∩ S₂) + S₂ * (S₁ ∩ S₂) : add_mul
+                    ... ⊆ S₁ * S₂ + S₂ * S₁ : add_subset_of_subset_of_subset
+                                                (mul_subset_of_subset_of_subset (λ x, id) (λ x hx, hx.2))
+                                                (mul_subset_of_subset_of_subset (λ x, id) (λ x hx, hx.1))
+                    ... = S₁ * S₂ + S₁ * S₂ : add_congr rfl mul_comm
+                    ... = S₁ * S₂ : add_self
+
+variables (S₁ S₂)
+
+def coprime : Prop := S₁ + S₂ = principal_ideal 1
+
+variables {S₁ S₂}
+
+theorem inter_eq_mul_of_coprime : coprime S₁ S₂ → S₁ ∩ S₂ = S₁ * S₂ :=
+λ h, set.eq_of_subset_of_subset
+(calc
+  S₁ ∩ S₂ = (univ_ideal α) * (S₁ ∩ S₂) : univ_mul.symm
+      ... = (principal_ideal 1) * (S₁ ∩ S₂) : mul_congr (principal_ideal_one_eq_univ α).symm rfl
+      ... = (S₁ + S₂) * (S₁ ∩ S₂) : mul_congr h.symm rfl
+      ... ⊆ S₁ * S₂ : add_mul_inter_subset_mul)
+mul_subset_inter
+
+end operations_on_ideals
+
+end is_ideal
+
+def tuple : list Type.{u} → Type u
```

Comment:
Why do you introduce these products? The indexed products should be much easier to use!

(Previous comment, I didn't see the later part: 
This is an impressive theory.
However, I think for rings we want to have a more powerful construction: Construction of arbitrary products:
```lean
instance comm_ring.Pi {α : Type u} {β : α → Type v} [∀a, comm_ring (β a)] : comm_ring (Πa, β a) :=
_
```
This does not enforce a order on the elements of our product, and it does not restrict the product construction to finite products. In this form however it is the box product, instead of the direct product. For the direct product we can use `finsupp` or better develop a dependent typed version of `finsupp`.)

Reply:
direct sum has finite support; direct product does not.